src/utils.js

```diff
@@ -108,7 +133,155 @@ async function createSequenceIfNotExists (sequenceName, actions, annotations, we
+  }
+  if (providers.length === 1) {
+    aioLogger.debug('There is a single matching event provider found for event')
+    return providers[0]
+  }
+  aioLogger.debug('Multiple event providers found for the event code. Initiating selection dialog...')
+  const message = 'We found multiple event providers for event type ' + eventType + '. Please select provider for this project'
+  const choices = providers.map(e => { return { name: e.label, value: e.id, instance_id: e.instance_id } })
+
+  const result = await prompt([
+    {
+      type: 'list',
+      name: 'res',
+      message,
+      choices
+    }
+  ])
+
+  return providers.find(e => e.id === result.res)
+}
+
+/**
+ * Setup I/O Events client and credentials
+ *
+ * Adds I/O Management permissions to current credentials if needed
+ *
+ * @param {*} project - project object
+ * @param {*} workspace - workspace object
+ * @param {*} options - hook's options
+ * @returns {*} - events client
+ */
+async function setupEventsClient (project, workspace, options) {
+  const projectConfig = coreConfig.get('project')
+  const orgId = projectConfig.org.id
+  const orgCode = projectConfig.org.ims_org_id
+  const env = getCliEnv()
+  await context.setCli({ 'cli.bare-output': true }, false) // set this globally
+  const accessToken = await getToken(CLI)
+  const cliObject = await context.getCli()
+  const apiKey = CONSOLE_API_KEYS[env]
+
+  const consoleCLI = await LibConsoleCLI.init({ accessToken: cliObject.access_token.token, env, apiKey: apiKey })
+  const workspaceCreds = await consoleCLI.getFirstEntpCredentials(orgId, project.id, workspace)
+  const client = await eventsSdk.init(orgCode, workspaceCreds.client_id, accessToken)
+
+  const supportedServices = await consoleCLI.getEnabledServicesForOrg(orgId)
+  const currentServiceProperties = await consoleCLI.getServicePropertiesFromWorkspace(
+    orgId,
+    project.id,
+    workspace,
+    supportedServices
+  )
+
+  const isIOManagementPresent = currentServiceProperties.filter(function (item) {
+    return item.sdkCode === 'AdobeIOManagementAPISDK'
+  })
+
+  if (isIOManagementPresent.length === 0) {
+    aioLogger.debug('IO Management API is not available to aio cli. Adding the service...')
+    currentServiceProperties.push({
+      name: 'I/O Management API',
+      sdkCode: 'AdobeIOManagementAPISDK',
+      roles: null,
+      licenseConfigs: null
+    })
+
+    await consoleCLI.subscribeToServices(
+      orgId,
+      project,
+      workspace,
+      path.join(options.config.dataDir, ENTP_INT_CERTS_FOLDER),
+      currentServiceProperties
+    )
+
+    const serviceConfig = currentServiceProperties.map(s => ({
+      name: s.name,
+      code: s.sdkCode
+    }))
+    coreConfig.set(AIO_CONFIG_WORKSPACE_SERVICES, serviceConfig, true)
+  } else {
+    aioLogger.debug('Current aio cli token allows access to IO management API')
+  }
+
+  return {
+    client,
+    workspaceCreds
+  }
+}
```

Comment:
Getting an event client seems to be a very common task as I am also doing this in the Commerce extensibility template. Should we be using this code snippet from some common library rather than repeating this across various plugins and templates?

Reply:
Looks like a canditate for extraction, though at the moment the only code using this logic seems to be this plugin. Here are organization search for this logic : https://github.com/search?p=2&q=org%3Aadobe+AdobeIOManagementAPISDK&type=Code